src/lib.rs

```diff
@@ -240,36 +236,36 @@ impl State {
 
     #[inline(always)]
     pub fn set_check(&mut self, x: u8) {
-        self.fach &= !CHECK_MASK;
-        self.fach |= u32::from(x);
+        self.opos_ch &= !CHECK_MASK;
+        self.opos_ch |= u32::from(x);
     }
 
     #[inline(always)]
     pub fn set_fail(&mut self, x: u32) {
-        self.fach &= !FAIL_MASK;
-        self.fach |= x << 8;
+        self.fail = x;
     }
 
     #[inline(always)]
     pub fn set_output_pos(&mut self, x: u32) {
-        self.output_pos = x;
+        self.opos_ch &= CHECK_MASK;
+        self.opos_ch |= x << 8;
     }
```

Comment:
I think `x` should be checked.
```rust
use crate::errors::{DaachorseError, Result};

pub fn set_output_pos(&mut self, x: u32) -> Result<()> {
    if ... {
        ...
        Ok(())
    } else {
        Err(DaachorseError::automaton_scale("outputs.len()", OUTPUT_POS_INVALID))
    }
}
```

Reply:
The upper bound of a given `x` can be checked by adding an error handling if `nfa.len >=  OUTPUT_POS_INVALID` at https://github.com/daac-tools/daachorse/blob/change-layout-state/src/builder.rs#L289-L291

So, I guess that the error handling in `set_output_pos` should be written with `debug_assert` for time overhead.

Reply:
Basically, the checking should not be separated from the function, so the checking code in the `build_sparse_nfa()` function should be moved into the `set_output_pos()` function.
If `set_output_pos()` does not check the argument, this function should be `unsafe`.

Reply:
I see and agree with that `set_output_pos` contains error handling.

> so the checking code in the build_sparse_nfa() function should be moved

But, I think that `build_sparse_nfa` has also to contain the error handling to quickly detect the scale error.

Reply:
The checking cost is not significant, so I think it is better to check that twice to be safe.

Reply:
updated in https://github.com/daac-tools/daachorse/pull/32/commits/8eef8e2b74c83f0e10455bede232371422495be7.